File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    
    phone_book.sort()
    length = len(phone_book)

    for i in range(length - 1):
        if len(phone_book[i]) <= len(phone_book[i + 1]):
            if phone_book[i + 1][:len(phone_book[i])] == phone_book[i]:
                return False
    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        /*
        phone_book 최대 길이 1,000,000이기에 O(N^2) 미만으로 해결
        
        [풀이]
        - 문자열 오름차순 정렬
        - 앞뒤로 접두사인지 확인
        */      
        Arrays.sort(phone_book); // O(NlogN)
        int num = phone_book.length;
        
        // System.out.println(Arrays.toString(phone_book));
        
        for (int i = 0; i < num - 1; i++) { // O(N)
            String cur = phone_book[i];
            String next = phone_book[i + 1];
            
            if (next.startsWith(cur)) return false;
        }

        return true;
    }
}